### PR TITLE
feat: added partial shell generation using root params

### PR DESCRIPTION
**Pull request by wyattjoh**
This enables the generation of partial shells when using with partial prerendering, aided by the new `rootParams()` API. Essentially, when your application only returns partial routes (where not all the route parameters are known), Next.js will now build route shells for these pages. We call these shells, fallback shells. They represent the partial state of the page that once served to the user, will provide a more complete loading experience as fast as possible.

We'll also take every permutation of the provided root params and generate a shell just for them to ensure that any future calls to those routes will be able to use the more specific fallback shell that's generated rather than having to rely on a blank shell by default.

### CODE REVIEW

**Comment by wyattjoh** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/next.js/73816?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#73816** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73816?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/73816?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#72837** <a href="https://app.graphite.dev/github/pr/vercel/next.js/72837?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 0708a6aef2a097d2d95397be723a14ab4e3ab86d

`pnpm test-start-turbo test/e2e/app-dir/app-root-params/multiple-roots.test.ts` (turbopack)
- app-root-params - multiple roots > should have root params on dashboard pages
- app-root-params - multiple roots > should not have root params on marketing pages

<details>
<summary>Expand output</summary>

  ● app-root-params - multiple roots › should have root params on dashboard pages

    next build failed with code/signal 1

       98 |           if (code || signal)
       99 |             reject(
    > 100 |               new Error(`next build failed with code/signal ${code || signal}`)
          |               ^
      101 |             )
      102 |           else resolve()
      103 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:100:15)

  ● app-root-params - multiple roots › should not have root params on marketing pages

    next build failed with code/signal 1

       98 |           if (code || signal)
       99 |             reject(
    > 100 |               new Error(`next build failed with code/signal ${code || signal}`)
          |               ^
      101 |             )
      102 |           else resolve()
      103 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:100:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| buildDuration | 20.1s | 18.5s | N/A |
| buildDurationCached | 17.5s | 14.8s | N/A |
| nodeModulesSize | 409 MB | 410 MB | ⚠️ +46.4 kB |
| nextStartRea..uration (ms) | 514ms | 518ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| 1187-HASH.js gzip | 50.7 kB | 50.7 kB | N/A |
| 8276.HASH.js gzip | 169 B | 168 B | N/A |
| 8377-HASH.js gzip | 5.36 kB | 5.36 kB | N/A |
| bccd1874-HASH.js gzip | 53 kB | 53 kB | N/A |
| framework-HASH.js gzip | 57.5 kB | 57.5 kB | N/A |
| main-app-HASH.js gzip | 231 B | 235 B | N/A |
| main-HASH.js gzip | 34 kB | 34 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 512 B | 510 B | N/A |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | ✓ |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 363 B | 362 B | N/A |
| hooks-HASH.js gzip | 393 B | 392 B | N/A |
| image-HASH.js gzip | 4.49 kB | 4.49 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.35 kB | 2.34 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 326 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.59 kB | 3.59 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| _buildManifest.js gzip | 747 B | 746 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| index.html gzip | 524 B | 524 B | ✓ |
| link.html gzip | 539 B | 539 B | ✓ |
| withRouter.html gzip | 520 B | 520 B | ✓ |
| Overall change | 1.58 kB | 1.58 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| edge-ssr.js gzip | 128 kB | 128 kB | N/A |
| page.js gzip | 203 kB | 203 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 672 B | 666 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 31.2 kB | 31.2 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 844 B | 844 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| 523-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 523.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 323 kB | 323 kB | ✓ |
| app-page-exp..prod.js gzip | 127 kB | 127 kB | ✓ |
| app-page-tur..prod.js gzip | 140 kB | 140 kB | ✓ |
| app-page-tur..prod.js gzip | 135 kB | 135 kB | ✓ |
| app-page.run...dev.js gzip | 313 kB | 313 kB | ✓ |
| app-page.run..prod.js gzip | 123 kB | 123 kB | ✓ |
| app-route-ex...dev.js gzip | 37.3 kB | 37.3 kB | ✓ |
| app-route-ex..prod.js gzip | 25.4 kB | 25.4 kB | ✓ |
| app-route-tu..prod.js gzip | 25.4 kB | 25.4 kB | ✓ |
| app-route-tu..prod.js gzip | 25.2 kB | 25.2 kB | ✓ |
| app-route.ru...dev.js gzip | 38.9 kB | 38.9 kB | ✓ |
| app-route.ru..prod.js gzip | 25.2 kB | 25.2 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.67 kB | 9.67 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.6 kB | 11.6 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.66 kB | 9.66 kB | ✓ |
| pages-turbo...prod.js gzip | 21.7 kB | 21.7 kB | ✓ |
| pages.runtim...dev.js gzip | 27.4 kB | 27.4 kB | ✓ |
| pages.runtim..prod.js gzip | 21.7 kB | 21.7 kB | ✓ |
| server.runti..prod.js gzip | 916 kB | 916 kB | ✓ |
| Overall change | 2.36 MB | 2.36 MB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js feat/generate-partial-shells | Change |
| - | - | - | - |
| 0.pack gzip | 2.05 MB | 2.05 MB | ⚠️ +3.35 kB |
| index.pack gzip | 72.5 kB | 71.1 kB | N/A |
| Overall change | 2.05 MB | 2.05 MB | ⚠️ +3.35 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: b0881dda54f4026da35821836237351c5d1a7316

File: packages/next/src/build/segment-config/app/collect-root-param-keys.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      rootParams.push(param)
+    }
+
+    // If this has a layout module, then we've found the root layout because
+    // we return once we found the first layout.
+    if (typeof modules.layout !== 'undefined') {
+      return rootParams
+    }
+
+    // This didn't include a root layout, so we need to continue.
```
Comment (wyattjoh) [inline packages/next/src/build/segment-config/app/collect-root-param-keys.ts:33]: Add documentation:

You can't have a parallel route before a root layout, so we only need to go through the `children` parallel route.

File: packages/next/src/build/static-paths/types.ts [hunk 2]
```diff
@@ -1,15 +1,19 @@
 import type { FallbackMode } from '../../lib/fallback'
 
 type StaticPrerenderedRoute = {
-  path: string
+  decoded: string
   encoded: string
```
Comment (wyattjoh) [inline packages/next/src/build/static-paths/types.ts:5]: Rename to:

```
route.pathname
route.decodedPathname
```

File: packages/next/src/build/static-paths/pages.ts [hunk 3]
```diff
@@ -159,51 +149,33 @@ export async function buildStaticPaths({
         ) {
           paramValue = []
         }
+
         if (
           (repeat && !Array.isArray(paramValue)) ||
-          (!repeat && typeof paramValue !== 'string')
+          (!repeat && typeof paramValue !== 'string') ||
+          !paramValue
```
Comment (wyattjoh) [inline packages/next/src/build/static-paths/pages.ts:156]: Pull this into a separate PR

File: packages/next/src/build/static-paths/app.ts [hunk 4]
```diff
@@ -256,36 +434,98 @@ export async function buildAppStaticPaths({
...
+            fallbackParamsIncludesRootParams
+            ? FallbackMode.BLOCKING_STATIC_RENDER
+            : fallbackMode
+          : FallbackMode.NOT_FOUND,
+        fallbackRootParams,
+      })
     })
   }
 
+  console.log('RESULT', JSON.stringify(result, null, 2))
```
Comment (wyattjoh) [inline packages/next/src/build/static-paths/app.ts:527]: Delete.

File: packages/next/src/build/index.ts [hunk 5]
```diff
@@ -2885,8 +2893,11 @@ export default async function build(
           }
 
           staticPaths.forEach((prerenderedRoutes, originalAppPath) => {
-            const page = appNormalizedPaths.get(originalAppPath) || ''
-            const appConfig = appDefaultConfigs.get(originalAppPath) || {}
+            const page = appNormalizedPaths.get(originalAppPath)
+            if (!page) throw new Error('Page not found')
```
Comment (wyattjoh) [inline packages/next/src/build/index.ts:2897]: Refactor: `InvariantError`

File: packages/next/errors.json [hunk 6]
```diff
@@ -613,5 +613,10 @@
...
   "614": "Route %s used \"unstable_rootParams\" inside \"use cache\". This is not currently supported.",
-  "615": "Missing workStore in unstable_rootParams"
-}
+  "615": "Missing workStore in unstable_rootParams",
+  "616": "App config not found",
+  "617": "A required parameter (%s) was not provided as a string received %s in generateStaticParams for %s",
+  "618": "A required parameter (%s) was not provided as an array received %s in generateStaticParams for %s",
+  "619": "Page not found",
+  "620": "A required parameter (%s) was not provided as %s received %s in getStaticPaths for %s"
+}
```
Comment (ztanner) [inline packages/next/errors.json:622]: I think this needs linting

File: packages/next/src/build/segment-config/app/collect-root-param-keys.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    // route above a root layout.
+    current = parallelRoutes.children
+  }
+
+  // If we didn't find a root layout, then we don't have any params.
+  return []
+}
+
+/**
+ * Collects the segments for a given route module.
```
Comment (ztanner) [inline packages/next/src/build/segment-config/app/collect-root-param-keys.ts:44]: looks like the comment here needs to be updated to reflect that this function is collecting root params

File: packages/next/src/build/static-paths/app.ts [hunk 8]
```diff
@@ -1,33 +1,240 @@
...
+      const item = unique[i]
+      let j = 0
+      for (; j < paramKeys.length; j++) {
+        const key = paramKeys[j]
+
+        // If the param is not the same, then we need to break out of the loop.
+        if (!areParamValuesEqual(item[key], params[key])) {
+          break
+        }
+      }
```
Comment (ztanner) [inline packages/next/src/build/static-paths/app.ts:76]: (non-blocking nit)

this function is a pretty clever usage of loops but I worry that what it makes up for in performance it sacrifices in readability. I'm also not 100% sure that this is more performant than the more expressive version, eg:

```js
 for (const params of routeParams) {
    const isDuplicate = unique.some((item) =>
      paramKeys.every((key) => areParamValuesEqual(item[key], params[key]))
    );

    if (!isDuplicate) {
      unique.push(params);
    }
  }
  ```
  
  In the worst case, this is O(N * M * K), which I think matches the implementation here as well:
  
  ```
  for each params in routeParams (N times):
    for each item in unique (M times):
      for each key in paramKeys (K times):
      // perform comparison
  ```
  
(I actually am curious if the built-in loops that power `some`/`every` might be slightly more efficient?)

**Review by ztanner — APPROVED** [review]

File: packages/next/src/build/static-paths/app.ts (hunk 8 above)
Comment (wyattjoh) [reply packages/next/src/build/static-paths/app.ts:76]: The only downside to the `.some.every` version is that each loop of `N`, we create 2 functions, which is avoided with the looping. I'll defer to @gnoff if this seems alright 😅

Comment (gnoff) [reply packages/next/src/build/static-paths/app.ts:76]: I'm partial to the loop code b/c it should be faster both in internal implementation and b/c it creates fewer functions. Typlically for loops can be heavily optimized by JIT and the builtins are only optimized maybe in the newest runtimes and I suspect not likely nearly as much as raw loops. That said I think you can achieve better complexity by tracking the param 'key' becuase the routeParams are in order and spaces aren't allowed.

for /app/[foo]/[bar]/page.tsx
we'd encode
`{}` -> `''`
`{ foo: 'f0' }` -> `'f0'`
`{ bar: 'b1',  foo: 'f1' }` -> `'f1 b1'`
`{ foo: 'f2', bar: 'b2' }` -> `'f2 b2'`
And so if you later encountered `{ bar: 'b2', foo: 'f2' }` (same values, different object key order) you'd just recompute `'f2 b2'` and see you already produced this result

It doesn't even have to be a string you could also have Maps of Maps but this has fewer objects created

The interesting question is what happens if you had just `{ bar: 'b3' }`. This would really be treated like unknown params from the foo downwards so it should encode as `''` since the bar can't be "reached" except if there was a known foo value.

**Comment by wyattjoh** [comment]
### Merge activity

* **Dec 17, 12:56 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/73816).
